### ISSUE TITLE
Weird Side Effects of `loadparams!`

### ISSUE THREAD

**Issue by FelixBenning** [FluxML/Flux.jl#1979]
I am posting a picture of the pluto notebook for better understanding/readability (what are the outputs etc). Below you can find copyable code.

<img width="740" alt="Screenshot 2022-05-30 at 16 29 49" src="https://user-images.githubusercontent.com/29106652/171013301-5015774a-15e5-4ae3-b9cc-ed51300ef9fd.png">

<details> <summary> Copyable Code</summary>

```julia
begin
	using Flux:Flux
	using MLDatasets: MNIST
	using LinearAlgebra: norm, LinearAlgebra
end

MNIST.download(i_accept_the_terms_of_use=true)

begin
	x_train, y_train = MNIST.traindata()
	x_train = Float32.(x_train)
	y_train_oh = Flux.onehotbatch(y_train, 0:9)
	size(x_train), size(y_train), size(y_train_oh)
end

function carth_grid(dim=3; start=0, stop=5, length=11)
	Iterators.product(fill(range(start, stop, length=length), dim)...)
end


function toy_model()
	return Flux.Chain(
		Flux.flatten,
		Flux.Dense(foldl(*, size(x_train[:,:,1])), 10, bias=false)
	)
end

begin
	grid = carth_grid(2, start=0, stop=1, length=2)
	collect(grid)
end

begin
	model = toy_model()
	origin = Flux.params(model)
	directions = map(1:length(size(grid))) do _
		dir = [x for x in Flux.params(toy_model())]
		return dir/norm(dir)
	end
	r_grid = map(grid) do coords
		ps = origin .+ sum(zip(coords, directions)) do (coeff, dir)
			coeff * dir
		end
		# Flux.loadparams!(model, deepcopy(ps))
		return ps
	end
	map(pairs(r_grid)) do (x,y)
		norm(x-y)
	end
end
```

</details>

Okay so what is happening? I define a generator of a model `toy_model`, to essentially generate different parameter initializations. Because apparently that is faster than using loadparams https://github.com/FluxML/Flux.jl/issues/1764 and I can use the default initializer to get a more realistic distribution over initializations. Then I sample two prameter vectors from this distribution and normalize them. Since they are random with high dimension, they are very likely to be almost orthogonal. So due to normalization we get orthonormal vectors. So when I create the grid using the coordinates from a carthesian grid, I am essentially doing an orthonormal basis change. 

Okay, so far so good. Now when I determine the distances of all the points on the new grid, the end result should be the same as before the basis change. Now the longest distance is between the points `(0,0)` and `(1,1)`. Which is the squareroot of 2. So ~1.4. And that is in fact the largets value of the output. Nice! Everything works as intended.

But if I comment in the `Flux.loadparams!` my new output makes no sense

```
[1.0, 1.41621, 2.83242, 1.0, 2.23859, 1.41621]
```

This implies that `loadparams!` somehow modifies ps even through the deepcopy?

**Comment by FelixBenning** [comment]
I guess this is a good example, why parameters of a model should not be implicit. Normalizing them etc. was also a pain in the ass, as you can not use `norm` on the `Flux.Params` object directly, so I had to do the `[x for x in Flux.params(...)]` thing. This entire thing would be so much easier if you wouldn't have to keep track of some form of hidden state. But tensorflow/pytorch aren't any better, and I want to use some built in elements.

**Comment by darsnack** [comment]
Not at the computer, but a quick suggestion would be to use `Flux.fmap` (from Functors.jl) to modify the parameters instead of dealing with `Params`. Similarly use `loadmodel!` instead of `loadparams!` (the latter is something we eventually want to deprecate). Both of these functions are the explicit alternatives to working with `Params`.

**Comment by FelixBenning** [comment]
Both fmap and loadmodel force a one to one relation between models and params. But I want to generate a grid of params and then try them out, plugging them into the model. So after I generated two directions of parameter vectors (by creating the model multiple times, which is already somewhat confusing), I use those directions to define a grid of parameters. Let us for example consider the point `(1,2)` This implies I want to add one times the first parameter direction, to two times the second parameter direction. Okay I can `fmap` the second parameter vector with `x->2*x`. But I still need to add them together. Now if they are both part of their models and I can't work with the parameters directly, that is impossible. `fmap` does not really allow me to `zip` two or more models (depending on the dimension of my grid). `loadmodel` is not helpful since I don't want to just dump/restore models to json/bson. I want to do arithmetic on the parameter vectors.

**Comment by darsnack** [comment]
Looks like Brian beat me to the punch, but I already typed this out, so here it is.

----

Okay let me see if I understand the problem statement correctly. You have a "model distribution" induced by the initialization functions. You want to take two samples from this distribution (or do you want to sample two parameters from a single initialization? I was confused on this point). Then we think of these two model initializations as a flat vector of parameters in parameter space. We normalize them to make them unit vectors. They are also extremely likely to be orthogonal, giving me an orthonormal basis. Then I define a Cartesian grid of coordinates, allowing me to explore linear combinations of my basis.

If the above is correct, then you could try the following.

### Explicit option

This one is admittedly more complex, mostly because we haven't exposed using Functors.jl except in the most basic cases. There's work to be done to make this more friendly for users.

```julia
using Flux
using Flux: fmap, destructure, trainable

# this would eventually be something a user does not define
function walk_trainable(f, x, ys...)
    func, re = trainable(x)
    yfuncs = map(trainable, ys)
    re(map(f, func, yfuncs...))
end

function normalize_model(m)
    flat, rebuild = destructure(m)
    
    return rebuild(flat / norm(flat))
end

function combine_models(coefficients, models)
    scaled_models = map(coefficients, models) do scale, m
        fmap(m; walk = walk_trainable) do p
            scale .* p
        end
    end
    combined_model = fmap(sum, scaled_models...; walk = walk_trainable)
    
    return combined_model
end

model_samples = [normalize_model(toy_model()) for _ in grid_cardinality]
grid = # make the grid

results = map(grid) do coefficients
    combine_models(coefficients, models)
end
```

Another option using `destructure` for everything.

```julia
using Flux
using Flux: destructure

function normalize_model(m)
    flat, rebuild = destructure(m)
    
    return rebuild(flat / norm(flat))
end

function combine_models(coefficients, models)
    flat, rebuild = destructure(models[1])
    flats = [destructure(m)[1] for m in models[2:end]]
    pushfirst!(flats, flat)
    combined_flats = mapreduce(coefficients, +, flats) do c, p
        c .* p
    end
    
    return rebuild(combined_flats)
end

model_samples = [normalize_model(toy_model()) for _ in grid_cardinality]
grid = # make the grid

results = map(grid) do coefficients
    combine_models(coefficients, models)
end
```

### Implicit option

```julia
using Flux
using Flux: Params, params

# even here it might be better to use destructure
# but I am sticking with implicit on purpose
function normalize_params!(model)
    ps = params(model)
    flat = mapreduce(vec, vcat, ps)
    flat .= flat ./ norm(flat)
    copy!(ps, flat)
    
    return model
end

function combine_models!(dst, coefficients, models)
    ps = params(dst)
    ps .= coefficients[1] .* params(models[1])
    for (c, m) in zip(coefficients, models[2:end])
        ps .= ps .+ c .* params(m)
    end
    
    return dst
end

model_samples = [normalize_params!(toy_model()) for _ in grid_cardinality]
grid = # make the grid

dst = toy_model()
results = map(grid) do coefficients
    combine_models!(dst, coefficients, models)
end
```

**Comment by FelixBenning** [comment]
Thank you for all the suggestions - I'll give it another shot tomorrow :)

**Comment by darsnack** [comment]
Most of the cost is walking the model which you have to do no matter what. The `rebuild` method of each model is cheap to throw away, I think. It only stores a vector of offsets and the total parameter length (see [here](https://github.com/FluxML/Optimisers.jl/blob/master/src/destructure.jl)). You do pay some cost, of course, but it should be nearly negligible compared to evaluating the model or taking gradients.

But the larger problem of assuming the models are the same structure...if they weren't, then the concept of linear combinations of them wouldn't make sense. You could add checks to `walk_trainable` to make sure the child nodes are the same. `loadmodel!` is one example where [we do this](https://github.com/FluxML/Flux.jl/blob/master/src/loading.jl), because you want to be very particular and conservative about loading one model into another. These kinds of checks should be built into `fmap` or whatever the user facing function is. Doing the checks with `destructure` is trickier cause you only walk a single model at a time.

**Comment by FelixBenning** [comment]
@ToucheSir I like the destructure/restructure mechanic actually. It is quite readable and side effect free judging by the lack of `!`. Using `loadmodel!` doesn't have that property, and also with fmap there would be at least another copy if I understand what you are suggesting (flatten the model, do the transformation, load back into model with fmap). I mean then you are using fmap like a restructure function. Just with less readability.

**Comment by ToucheSir** [comment]
It's exactly the opposite: `destructure` _may_ induce a copy for some parameters (it definitely will the first time around), whereas handling the nested structure using ` fmap` and co. are guaranteed not to add any extra ones (modulo what you yourself allocate in the callback).
